src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1346,4 +1368,149 @@ public static MethodHandle makeConcatWithTemplateGetters(
+        static final MethodTypeDesc APPEND_INT_TYPE = MethodTypeDesc.ofDescriptor("(I)Ljava/lang/StringBuilder;");
+        static final MethodTypeDesc APPEND_LONG_TYPE = MethodTypeDesc.ofDescriptor("(J)Ljava/lang/StringBuilder;");
+        static final MethodTypeDesc APPEND_OBJECT_TYPE = MethodTypeDesc.ofDescriptor("(Ljava/lang/Object;)Ljava/lang/StringBuilder;");
+        static final MethodTypeDesc APPEND_STRING_TYPE = MethodTypeDesc.ofDescriptor("(Ljava/lang/String;)Ljava/lang/StringBuilder;");
+
+        static {
+            DUMPER = ClassFileDumper.getInstance("java.lang.invoke.StringConcatFactory.dump", "stringConcatClasses");
+        }
+
+        private SimpleStringBuilderStrategy() {
+            // no instantiation
+        }
+
+        private static MethodHandle generate(Lookup lookup, MethodType args, String[] constants) throws Exception {
+            String className = getClassName(lookup.lookupClass());
+
+            byte[] classBytes = ClassFile.of().build(ClassDesc.of(className),
+                    new Consumer<ClassBuilder>() {
+                        @Override
+                        public void accept(ClassBuilder clb) {
+                            clb.withFlags(AccessFlag.PUBLIC, AccessFlag.FINAL, AccessFlag.SUPER, AccessFlag.SYNTHETIC)
+                                .withMethodBody(METHOD_NAME,
+                                        MethodTypeDesc.ofDescriptor(args.toMethodDescriptorString()),
+                                        ClassFile.ACC_FINAL | ClassFile.ACC_PUBLIC | ClassFile.ACC_STATIC,
+                                        generateMethod(constants, args));
+                    }});
+
+            try {
+                Class<?> innerClass = lookup.defineHiddenClass(classBytes, true, STRONG).lookupClass();
+                DUMPER.dumpClass(className, innerClass, classBytes);
+                MethodHandle mh = lookup.findStatic(innerClass, METHOD_NAME, args);
+                return mh;
+            } catch (Exception e) {
+                DUMPER.dumpFailedClass(className, classBytes);
+                throw new StringConcatException("Exception while spinning the class", e);
+            }
+        }
+
+        private static Consumer<CodeBuilder> generateMethod(String[] constants, MethodType args) {
+            return new Consumer<CodeBuilder>() {
+                @Override
+                public void accept(CodeBuilder cb) {
+                    cb.new_(STRING_BUILDER);
+                    cb.dup();
+                    cb.invokespecial(STRING_BUILDER, "<init>", MethodTypeDesc.ofDescriptor("()V"));
+
+                    // At this point, we have a blank StringBuilder on stack, fill it in with .append calls.
+                    {
+                        int off = 0;
+                        for (int c = 0; c < args.parameterCount(); c++) {
+                            if (constants[c] != null) {
+                                cb.ldc(constants[c]);
```

Comment:
I think for Remi's approach, we change:
1. Insert an extra String array (maybe need a way to mark it stable?) arg representing constants
2. Change this ldc into aload + aaload
3. Call `bindTo(constantStrings)` on the resulting MH

This approach can significantly reduce the number of classes spinned instead of generating one class per constant array; might need to measure performance to see if this is a good tradeoff

Reply:
I'd prefer considering such optimizations as follow-ups. We need to think about where to define such shared classes in a way that considers the full lifecycle, facilitates class unloading (one cache per classloader?) while still getting good reuse.

Reply:
I think `Invokers` might be a good place. Its lifetime is bound to its MT, and MT already handles class unloading well. If we are doing erased then a plain array suffices.

Reply:
I think we should be wary about adding things that are strongly kept alive by the MT. While the MTs themselves are weakly referenced and can be unloaded if all usage goes aways, many oft-used MTs might be kept effectively forever.